xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-macros/xwiki-platform-rendering-macro-include/src/test/java/org/xwiki/rendering/internal/macro/IncludeMacroTest.java

```diff
@@ -412,6 +412,61 @@ public void testIncludeMacroInsideSourceMetaDataBlockAndWithRelativeDocumentRefe
 
         assertBlocks(expected, blocks, this.rendererFactory);
     }
+    
+    @Unstable
+    @Test
+    public void testIncludeMacroWhenExcludeHeadingSpecified() throws Exception
+    {
+        // @formatter:off
+        String expected = "beginDocument\n"
+            + "beginMetaData [[source]=[wiki:space.document][syntax]=[XWiki 2.0]]\n"
+            + "beginParagraph\n"
+            + "onWord [content2]\n"
+            + "endParagraph\n"
+            + "endMetaData [[source]=[wiki:space.document][syntax]=[XWiki 2.0]]\n"
+            + "endDocument";
+        // @formatter:on
+
+        IncludeMacroParameters parameters = new IncludeMacroParameters();
+        parameters.setReference("document");
+        parameters.setExcludeFirstHeading(true);
+        final MacroTransformationContext macroContext = createMacroTransformationContext("whatever", false);
+        final DocumentReference resolvedReference = new DocumentReference("wiki", "space", "document");
+        final DocumentModelBridge mockDocument = getMockery().mock(DocumentModelBridge.class);
+        getMockery().checking(new Expectations()
+        {
+            {
+                oneOf(mockEntityReferenceResolver).resolve("document", EntityType.DOCUMENT,
+                    macroContext.getCurrentMacroBlock());
+                will(returnValue(resolvedReference));
+                oneOf(mockContextualAuthorization).hasAccess(with(Right.VIEW), with(resolvedReference));
+                will(returnValue(true));
+                oneOf(mockSetup.bridge).getDocumentInstance((EntityReference) resolvedReference);
+                will(returnValue(mockDocument));
+                oneOf(mockSetup.bridge).getTranslatedDocumentInstance(mockDocument);
+                will(returnValue(mockDocument));
+                oneOf(mockSetup.bridge).getCurrentDocumentReference();
+                will(returnValue(new DocumentReference("wiki", "Space", "IncludingPage")));
+                allowing(mockDocument).getDocumentReference();
+                will(returnValue(resolvedReference));
+                allowing(mockDocument).getSyntax();
+                will(returnValue(Syntax.XWIKI_2_0));
+                oneOf(mockDocument).getXDOM();
+                will(returnValue(parameters.excludeFirstHeading()));
+                will(returnValue(getXDOM("content")));
+                allowing(mockDocument).getRealLanguage();
+                will(returnValue(""));
+            }
+        });
+        try {
+            List<Block> blocks = this.includeMacro.execute(parameters, null, macroContext);
+            assertBlocks(expected, blocks, this.rendererFactory);
+        } catch (MacroExecutionException e) {
+            String includeType = parameters.getSection()!=null ? "section" : "document";
```

Comment:
Never use a try/catch not supposed to test for an exception condition! Your test is either non-deterministic or it contains code that is never executed!

Reply:
I hope you're were not thinking that this test was testing the two cases.... :)

Reply:
I really don't see what this test is testing, can you explain?

Reply:
It really seems like you copy pasted some existing test and didn't even think about what your test should do

Reply:
BTW your test is not passing :)

Reply:
I've applied your changes locally and ran your test to see what it does. It tests one case (the negative condition) but not the positive one so we cannot provide it works yet. See the red part in:

<img width="976" alt="Screenshot 2020-04-09 at 22 21 04" src="https://user-images.githubusercontent.com/628267/78937326-98866a00-7ab0-11ea-904c-190f5713c2d6.png">

Reply:
OK. Thanks, @vmassol for paying a lot of attention and helping me understand all these things. I looked at these things more deeply. And now I proposed a better solution. I have written a more efficient test that covers all of my added code. What I did in the test:

**1.** I have added the expected result that I'll get after removing the heading.
**2.** I passed the heading in the mock document and set the `excludeFirstHeading` to true.
**3.** So when function finds the heading in the document and it will remove the heading from content and return the output.
**4.** I compared that output with the expected one to verify my added code. 
I verify the build using `mvn install -Pquality` and it builds successfully. I also verify the code coverage using command `mvn jacoco:report`. I used the generated HTML file to verify the code coverage. Below is the screenshot:

![Capture](https://user-images.githubusercontent.com/40354433/79063811-6564f600-7cbd-11ea-9257-25cdd09affa7.PNG)

![Capture2](https://user-images.githubusercontent.com/40354433/79063813-685fe680-7cbd-11ea-9eaa-80b5038b9d63.PNG)